### ISSUE TITLE
improve: promptのトークン数調整

### ISSUE THREAD

**Issue by kyonenya** [kyonenya/next-langchain-pdf#2]
- [improve: getRelevantDocumentsで7件くらい取っておいて文字数が少なかったら足す · next-langchain-pdf](https://github.com/users/kyonenya/projects/5/views/1?pane=issue&itemId=33819880)
- [x] [feat: PDFが日本語かどうかでチャンクサイズを出し分け · next-langchain-pdf](https://github.com/users/kyonenya/projects/5/views/1?pane=issue&itemId=33762946)

## トークン数制限

text-davinci-003の最大トークン数制限は4096。これをプロンプトとレスポンスにどう割り振るか。

- promptTemplate: 46
- chunkDocument: 2,704
  - (3000 characters × 5)
- response: 524
- TOTAL: 3254

3000文字／チャンクでもエラーになった
```
error Error: Response error: {
  "error": {
    "message": "This model's maximum context length is 4097 tokens, however you requested 4129 tokens (3105 in your prompt; 1024 for the completion). Please reduce your prompt; or completion length.",
    "type": "invalid_request_error",
    "param": null,
    "code": null
  }
}
```

### トークン数計測

[Tokenizer - OpenAI Platform](https://platform.openai.com/tokenizer)

#### 2048文字／チャンク

Tokens
2,272
Characters
6874

#### 3000文字／チャンク

Tokens
2,704
Characters
9370

#### 4096文字／チャンク

Tokens
3,750
Characters
15684

**Comment by kyonenya** [comment]
ChatGPT-3.5-turboで実験。

### 524 char/chunk

There are several factors that contributed to the increase in P&G's net sales and organic sales.

1. Product innovation: P&G has a strong focus on innovation and continuously introduces new and improved products to meet consumer demands. This helps drive sales growth as consumers are attracted to new and innovative offerings.

2. Strategic acquisitions: P&G has made strategic acquisitions to expand its product portfolio and enter new markets. These acquisitions have contributed to the growth of net sales as they bring in additional revenue from new brands and categories.

3. Marketing and advertising: P&G invests heavily in marketing and advertising to promote its products and build brand awareness. Effective marketing campaigns help drive consumer demand and contribute to increased sales.

4. E-commerce growth: P&G has been actively investing in e-commerce capabilities and expanding its presence in online channels. With the increasing popularity of online shopping, this has helped drive net sales as more consumers choose to purchase P&G products online.

5. International expansion: P&G has a global footprint and continues to expand its presence in emerging markets. This allows the company to access new customer bases and tap into growing markets, resulting in increased net sales.

6. Cost-saving measures: P&G has implemented various cost-saving measures to improve efficiency and profitability. This includes streamlining operations, optimizing supply chain management, and reducing overhead costs. These measures contribute to higher net sales by improving the company's overall financial performance.

Overall, a combination of product innovation, strategic acquisitions, effective marketing, e-commerce growth, international expansion, and cost-saving measures have contributed to the increase in P&G's net sales and organic sales.

### 2048 char/chunk

There are several factors that contributed to the increase in P&G's net sales and organic sales. These factors include:

1. Product innovation: P&G continuously invests in research and development to develop and launch innovative products. New and improved products attract consumers and drive sales growth.

2. Marketing and advertising: P&G is known for its strong marketing and advertising campaigns that effectively communicate the benefits of its products to consumers. These campaigns help to increase brand awareness and drive sales.

3. Expansion into emerging markets: P&G has been expanding its presence in emerging markets such as China, India, and Brazil. These markets offer significant growth opportunities due to their large populations and rising middle class. As P&G continues to penetrate these markets, it experiences increased sales.

4. E-commerce growth: The growth of e-commerce has provided P&G with new avenues for sales. P&G has been investing in its e-commerce capabilities and partnering with online retailers to drive online sales growth.

5. Cost savings initiatives: P&G has implemented cost-saving initiatives to improve its profitability. These initiatives include streamlining its supply chain, reducing operational costs, and optimizing its manufacturing processes. As a result, P&G has been able to improve its margins and drive sales growth.

6. Consumer demand during the COVID-19 pandemic: The COVID-19 pandemic has led to increased demand for certain P&G products, such as cleaning and hygiene products. As consumers focused on cleanliness and hygiene during the pandemic, P&G experienced a surge in sales in these product categories.

Overall, a combination of product innovation, effective marketing, expansion into emerging markets, e-commerce growth, cost savings initiatives, and increased demand during the COVID-19 pandemic have contributed to the increase in P&G's net sales and organic sales.

### 3000 char/chunk

There are several factors that contributed to the increase in P&G's net sales and organic sales. 

1. Innovation and new product launches: P&G is known for its focus on innovation and regularly introduces new and improved products to the market. This helps to drive sales as consumers are attracted to these novel offerings.

2. Strong brand portfolio: P&G has a diverse portfolio of well-established and trusted brands in various categories such as beauty, grooming, and household care. These brands have a strong presence in the market and enjoy customer loyalty, which contributes to increased sales.

3. Marketing and advertising efforts: P&G invests heavily in marketing and advertising to promote its products and create brand awareness. Effective marketing campaigns help to drive sales and attract new customers.

4. Expansion into emerging markets: P&G has been expanding its presence in emerging markets such as China, India, and Brazil, where there is a growing population and increasing disposable income. This expansion has helped to boost sales and increase market share.

5. E-commerce growth: The rise in online shopping has provided P&G with new avenues for sales growth. P&G has invested in e-commerce platforms and developed digital marketing strategies to reach and engage with consumers online.

6. The impact of the COVID-19 pandemic: The COVID-19 pandemic has led to increased demand for certain P&G products, such as cleaning and hygiene products. The increased focus on personal hygiene and cleanliness has resulted in higher sales for these product categories.

Overall, a combination of factors such as innovation, strong brands, effective marketing, expansion into emerging markets, e-commerce growth, and the impact of the COVID-19 pandemic have contributed to the increase in P&G's net sales and organic sales.

### 4096 char/chunk

There are several factors that contributed to the increase in P&G's net sales and organic sales. 

1. Product innovation: P&G has invested heavily in developing innovative and high-quality products that appeal to consumers. These new products, such as Tide Pods or Gillette Fusion razors, have helped drive sales growth.

2. Expansion into emerging markets: P&G has focused on expanding its presence in emerging markets such as China, Brazil, and India. These markets have experienced rapid economic growth and rising middle-class populations, creating opportunities for increased sales.

3. Effective marketing and advertising: P&G has implemented successful marketing and advertising campaigns to promote its products. These campaigns, often featuring popular celebrities or influencers, have helped raise brand awareness and drive sales.

4. Strategic acquisitions: P&G has made several strategic acquisitions over the years to diversify its product portfolio and expand into new business areas. These acquisitions, such as the acquisition of Gillette in 2005, have contributed to sales growth.

5. E-commerce growth: P&G has adapted to the increasing preference for online shopping by expanding its e-commerce capabilities. This has allowed the company to reach a wider customer base and increase sales.

Overall, a combination of product innovation, expansion into emerging markets, effective marketing, strategic acquisitions, and e-commerce growth have contributed to the increase in P&G's net sales and organic sales.

### 524文字／チャンク

P&Gの純売上高と既存事業売上高の増加にはいくつかの要因がある。

製品の革新： P&Gはイノベーションに力を入れており、消費者の需要に応えるため、継続的に新製品や改良品を発表している。これは、消費者が新しく革新的な製品に魅力を感じるため、売上高の成長を促進するのに役立っている。

戦略的買収： P&Gは、製品ポートフォリオを拡大し、新市場に参入するために戦略的買収を行ってきた。これらの買収は、新しいブランドやカテゴリーからの追加収益をもたらし、売上高の成長に貢献している。

マーケティングと広告： P&Gは、製品の販売促進とブランド認知度向上のため、マーケティングと広告に多額の投資を行っている。効果的なマーケティング・キャンペーンは消費者の需要を喚起し、売上増に貢献している。

Eコマースの成長： P&GはEコマースへの投資を積極的に行い、オンライン・チャネルでのプレゼンスを拡大している。オンラインショッピングの普及に伴い、P&G製品をオンラインで購入する消費者が増えており、売上高の増加に貢献している。

国際的拡大： P&Gはグローバルに事業を展開しており、新興市場でのプレゼンス拡大を続けている。これにより、新たな顧客層へのアクセスや成長市場への参入が可能になり、売上高の増加につながった。

コスト削減策： P&Gは効率性と収益性を向上させるため、様々なコスト削減策を実施してきた。これには、業務の合理化、サプライチェーン・マネジメントの最適化、諸経費の削減などが含まれる。これらの施策は、会社全体の財務実績を改善することで、売上高の増加に貢献している。

全体として、製品の革新、戦略的買収、効果的なマーケティング、eコマースの成長、国際的な事業展開、コスト削減策の組み合わせが、P&Gの純売上高と既存事業の売上高の増加に寄与している。

### 2048文字／チャンク

P&Gの純売上高と既存事業売上高の増加にはいくつかの要因がある。これらの要因には以下が含まれる：

1. 製品の革新： P&Gは継続的に研究開発に投資し、革新的な製品を開発・発売している。新しく改良された製品は消費者を魅了し、売上高の成長を促進する。

2. マーケティングと広告： P&Gは、製品の利点を消費者に効果的に伝える強力なマーケティングと広告キャンペーンで知られている。これらのキャンペーンは、ブランド認知度を高め、売上を促進するのに役立っている。

3. 新興市場への進出： P&Gは、中国、インド、ブラジルなどの新興市場でのプレゼンスを拡大している。これらの市場は人口が多く、中間層が増加しているため、大きな成長機会がある。P&Gがこれらの市場に浸透し続けることで、売上が増加する。

4. Eコマースの成長： Eコマースの成長は、P&Gに新たな販売手段を提供した。P&Gはeコマース機能に投資し、オンライン小売業者と提携することで、オンラインでの売上拡大を推進している。

5. コスト削減の取り組み： P&Gは収益性を改善するため、コスト削減の取り組みを実施してきた。これらの取り組みには、サプライチェーンの合理化、業務コストの削減、製造プロセスの最適化などが含まれる。その結果、P&Gは利益率を改善し、売上成長を促進することができた。

6. COVID-19パンデミック時の消費者需要： COVID-19の大流行により、洗浄剤や衛生用品など特定のP&G製品に対する需要が増加した。パンデミックの間、消費者が清潔と衛生に注目したため、P&Gはこれらの製品カテゴリーの売上が急増した。

全体として、製品の革新、効果的なマーケティング、新興市場への進出、eコマースの成長、コスト削減の取り組み、COVID-19パンデミック時の需要増が組み合わさり、P&Gの純売上高と既存事業売上高の増加に貢献した。

### 3000文字／チャンク

P&Gの純売上高と既存事業売上高の増加にはいくつかの要因がある。

イノベーションと新製品の発売： P&Gはイノベーションに注力していることで知られ、定期的に新製品や改良品を市場に投入している。これは、消費者がこれらの斬新な製品に魅了され、売上を促進するのに役立っている。

強力なブランド・ポートフォリオ： P&Gは、美容、グルーミング、家庭用ケアなど様々なカテゴリーにおいて、定評と信頼のあるブランドの多様なポートフォリオを有している。これらのブランドは市場で強い存在感を示し、顧客ロイヤルティを獲得しているため、売上増に貢献している。

マーケティングと広告活動： P&Gは、製品の販売促進とブランド認知度向上のため、マーケティングと広告に多額の投資を行っている。効果的なマーケティング・キャンペーンは、売上促進と新規顧客の獲得に役立っている。

新興市場への進出： P&Gは、人口が増加し、可処分所得が増加している中国、インド、ブラジルなどの新興市場でのプレゼンスを拡大している。この拡大が売上を押し上げ、市場シェアの拡大に貢献している。

Eコマースの成長： オンライン・ショッピングの増加は、P&Gに売上拡大の新たな手段を提供した。P&Gはeコマース・プラットフォームに投資し、オンラインで消費者にリーチし、エンゲージするためのデジタル・マーケティング戦略を展開している。

COVID-19パンデミックの影響： COVID-19の大流行により、クリーニングや衛生用品など特定のP&G製品に対する需要が増加した。個人の衛生や清潔さへの関心が高まった結果、これらの製品カテゴリーの売上が増加した。

全体として、技術革新、強力なブランド、効果的なマーケティング、新興市場への進出、eコマースの成長、COVID-19パンデミックの影響などの複合的な要因が、P&Gの純売上高および既存事業売上高の増加に寄与した。

### 4096文字／チャンク

P&Gの純売上高と既存事業売上高の増加にはいくつかの要因がある。

1. 製品の革新： P&Gは、消費者にアピールする革新的で高品質な製品の開発に多額の投資を行ってきた。タイドポッドやジレット・フュージョンのカミソリなど、こうした新製品が売上高の伸びを後押ししている。

2. 新興市場への進出： P&Gは、中国、ブラジル、インドなどの新興市場でのプレゼンス拡大に注力してきた。これらの市場は急速な経済成長を遂げ、中産階級の人口が増加しており、売上増加の機会を生み出している。

3. 効果的なマーケティングと広告： P&Gは、製品の販売促進のためにマーケティングと広告キャンペーンを実施し、成功を収めてきた。これらのキャンペーンは、人気のある有名人やインフルエンサーを起用することが多く、ブランドの認知度を高め、売上を促進するのに役立っている。

4. 戦略的買収： P&Gは、製品ポートフォリオを多様化し、新たな事業分野に進出するため、長年にわたり戦略的買収を繰り返してきた。2005年のジレット買収など、こうした買収は売上成長に貢献している。

5. Eコマースの成長： P&Gは、eコマース機能を拡大することで、オンラインショッピングへの嗜好の高まりに適応してきた。これにより、同社はより幅広い顧客層にリーチし、売上を伸ばすことができた。

全体として、製品の革新、新興市場への進出、効果的なマーケティング、戦略的買収、Eコマースの成長などの組み合わせが、P&Gの純売上高と既存事業の売上高の増加に寄与している。

**Comment by kyonenya** [comment]
GPT-3は、非常に限られたデータセットや、偏った情報や誤解を招くような情報を持つデータセットでは苦戦するかもしれない。GPT-3は大量のデータからパターンを学習することに依存しているため、多様な、あるいは関連性のあるトレーニングデータが不足していると、うまく機能しない可能性がある。さらに、GPT-3は、ドメイン固有の知識や微妙な文脈の理解を必要とするタスクに苦戦する可能性がある。さらに、人間のような直感や感情を必要とする、非常に創造的で主観的なコンテンツを生成することも難しいかもしれません。

チャンク文字数
4096
Tokens
3,750
Characters
15684

**Comment by kyonenya** [comment]
## [ChatPDF](https://www.chatpdf.com/)

> Welcome to this exciting PDF file that explores the incredible capabilities of language models! In this document, you will discover how scaling up language models can greatly improve their few-shot performance, sometimes even reaching competitiveness with prior state-of-the-art fine-tuning approaches. Here are three example questions you might have about this file:

> 1.  What is a few-shot learner, and how does it differ from traditional machine learning approaches?
> 1. How does GPT-3, the autoregressive language model with 175 billion parameters, perform in the few-shot setting compared to other models?
> 1. What are some potential applications of few-shot learning in natural language processing, and how might it impact the field in the future?

#### 0. On which datasets does GPT-3 struggle?

> According to p. 5 and p. 18, GPT-3 struggles with some tasks, including natural language inference tasks like the ANLI dataset, some reading comprehension datasets like RACE or QuAC, and some comparison tasks where it does little better than chance when evaluated one-shot or even few-shot. P. 18 provides more specific information on GPT-3's performance on different datasets, noting that GPT-3 performs best on CoQA, a free-form conversational dataset, and worst on QuAC, a dataset which requires modeling structured dialog acts and answer span selections of teacher-student interactions. On DROP, a dataset testing discrete reasoning and numeracy in the context of reading comprehension, GPT-3 in a few-shot setting outperforms the fine-tuned BERT baseline from the original paper but is still well below both human performance and state-of-the-art approaches which augment neural networks with symbolic systems.

#### 1. What is a few-shot learner, and how does it differ from traditional machine learning approaches?

> According to 6, few-shot learning involves learning based on a broad distribution of tasks (in this case implicit in the pre-training data) and then rapidly adapting to a new task. Few-shot learning is related to few-shot learning as used in other contexts in ML, and both involve adapting to a new task with limited data. In contrast to traditional machine learning approaches, few-shot learning requires less task-specific data and can adapt to new tasks more quickly.

#### 2. How does GPT-3, the autoregressive language model with 175 billion parameters, perform in the few-shot setting compared to other models?

> p. 5 and p. 5 both discuss GPT-3's performance in the few-shot setting. According to p. 5, GPT-3 displays one-shot and few-shot proficiency at tasks designed to test rapid adaptation or on-the-fly reasoning, including unscrambling words, performing arithmetic, and using novel words in a sentence after seeing them defined only once. Additionally, GPT-3 can generate synthetic news articles in the few-shot setting which human evaluators have difficulty distinguishing from human-generated articles. However, p. 5 also notes that GPT-3 struggles with some tasks, including natural language inference tasks like the ANLI dataset and some reading comprehension datasets like RACE or QuAC. P. 33 further confirms that GPT-3's in-context learning performance has some notable gaps on a suite of benchmarks, including some comparison tasks where it does little better than chance when evaluated one-shot or even few-shot. Overall, GPT-3's few-shot performance is impressive in some areas but still has limitations.

#### 3. What are some potential applications of few-shot learning in natural language processing, and how might it impact the field in the future? 

> The PDF file does not provide a comprehensive list of potential applications of few-shot learning in natural language processing, but p. 5 suggests that few-shot learning could be useful for tasks that require rapid adaptation or on-the-fly reasoning, such as unscrambling words, performing arithmetic, and using novel words in a sentence after seeing them defined only once. Additionally, p. 33 notes that few-shot learning has been shown to be effective for fine-tuning language models on specific tasks, and that scaling up language models can greatly improve their few-shot performance. In the future, few-shot learning could potentially enable more efficient and effective natural language processing, as it requires less task-specific data and can adapt to new tasks more quickly than traditional machine learning approaches.

**Comment by kyonenya** [comment]
## 自作アプリ

#### On which datasets does GPT-3 struggle?

> GPT-3 struggles on datasets that require complex tasks such as natural language inference, reading comprehension, and discrete reasoning. Examples include the ANLI dataset, the QuAC dataset, the RACE dataset, and the DROP dataset. It can also struggle on SuperGLUE tasks such as RTE and NLI, as well as tasks that contain structured dialogues such as the QuAC dataset and the DROP dataset.

#### 1. What is a few-shot learner, and how does it differ from traditional machine learning approaches?

> A few-shot learner is a type of machine learning model that can learn from only a few examples or a simple instruction without the need for large datasets or gradient updates. This approach to learning differs from traditional machine learning models which generally require large datasets with labeled examples, as well as gradient updates. Few-shot learning is related to other types of few-shot learning which require a broad distribution of tasks and the ability to quickly adapt to a new task. In this paper, we train a 175 billion parameter language model, GPT-3, and measure its performance in the few-shot setting - demonstrating that larger models are more proficient at learning with fewer demonstrations. GPT-3 has achieved strong performance on many NLP datasets, including translation, question-answering, and cloze tasks, as well as several tasks that require domain adaptation.

#### 2. How does GPT-3, the autoregressive language model with 175 billion parameters, perform in the few-shot setting compared to other models?

> GPT-3 is able to perform well in the few-shot setting, often reaching competitiveness with existing state-of-the-art models that use fine-tuning. GPT-3 can achieve strong performance on many NLP datasets, including translation, question-answering, and cloze tasks. It is also capable of tasks that require on-the-fly reasoning or domain adaptation, such as unscrambling words, using a novel word in a sentence, or performing 3-digit arithmetic in the few-shot setting. However, GPT-3 may struggle with certain tasks, such as those that require comparison, common sense physics questions, or some reading comprehension tasks. GPT-3's few-shot learning performance improves with model size, and its performance on translation tasks has been aided by increased capacity and inclusion of other languages in the training data.

#### 3. What are some potential applications of few-shot learning in natural language processing, and how might it impact the field in the future?

> Few-shot learning has potential applications in natural language processing for tasks such as translation, question-answering, cloze tasks, and on-the-fly reasoning or domain adaptation, such as unscrambling words, using a novel word in a sentence, or performing 3-digit arithmetic. In addition, few-shot learning can also be used to correct English grammar, and to measure and prevent memorization of benchmarks. By leveraging few-shot learning methods, natural language processing models could perform tasks more quickly and be trained on a much smaller data set compared to traditional methods. This could result in more efficient and accurate natural language processing models with greater interpretability and computational efficiency. In the long term, few-shot learning could be a valuable tool in the development of larger and more powerful models with the ability to adapt quickly and effectively to changing conditions.

**Comment by kyonenya** [comment]
すごい。（chunkSize: 1000, chunkOverlap: 100, limit: 5）

![image](https://github.com/kyonenya/next-langchain-pdf/assets/62150154/1ffff686-c608-40ac-b1bb-74314a9cbf74)

![image](https://github.com/kyonenya/next-langchain-pdf/assets/62150154/9ce44398-a39d-4549-8acb-c2d8c1ec09ee)

**Comment by kyonenya** [comment]
chatgpt-3.5-turbo-16kモデル（最大トーク数：16385トークン）を使うことで力技で解決
トークン節約やめた
